<a href="https://colab.research.google.com/github/Carine-69/SchoolDropOut-RW/blob/main/Assignment1_CarineUmugabekazi.ipynb" target="_parent"><img src="https://colab.research.google.com/assets/colab-badge.svg" alt="Open In Colab"/></a>

**School Drop Out prediction**

1. INTRODUCTION
I this notebook, we predict whether a student will drop out, enroll, or graduate using a dataset of tudent records. We will use:
  1. logistic regression
  2.Decision Tree classifier
  3.Custom Neural Network(Three-lyer NN)

  We will preprocess the data, handle missing values, split it into training/testing, train models, and evaluate them.

  Dataset citation: Realinho, Valentim, Mónica Vieira Martins, Jorge Machado, and Luís Baptista. 2021. Predict Students' Dropout and Academic Success. UCI Machine Learning Repository. https://doi.org/10.24432/C5MC89.

In [67]:
# import libraries
import sklearn as sk
from sklearn.preprocessing import StandardScaler
from sklearn.model_selection import train_test_split, GridSearchCV
from sklearn.linear_model import LogisticRegression
from sklearn.tree import DecisionTreeClassifier
from sklearn.metrics import mean_squared_error, r2_score,classification_report, accuracy_score, confusion_matrix
from sklearn.pipeline import Pipeline
from sklearn.impute import SimpleImputer
import numpy as np
import matplotlib.pyplot as plt
import pandas as pd

Load and Explore Dataset

In [68]:
# connect to drive
from google.colab import drive
drive.mount('/content/drive')


Drive already mounted at /content/drive; to attempt to forcibly remount, call drive.mount("/content/drive", force_remount=True).


In [69]:
#import dataet
df = pd.read_csv('/content/drive/MyDrive/ML-datasets/SchoolDropOut.csv')

In [70]:
df.head()

Marital status  Application mode  Application order  Course  \
0               1                17                  5     171   
1               1                15                  1    9254   
2               1                 1                  5    9070   
3               1                17                  2    9773   
4               2                39                  1    8014   

   Daytime/evening attendance\t  Previous qualification  \
0                           1.0                       1   
1                           1.0                       1   
2                           1.0                       1   
3                           1.0                       1   
4                           0.0                       1   

   Previous qualification (grade)  Nacionality  Mother's qualification  \
0                           122.0            1                      19   
1                           160.0            1                       1   
2                           122.0            1                      37   
3                           122.0            1                      38   
4                           100.0            1                      37   

   Father's qualification  ...  Curricular units 2nd sem (credited)  \
0                      12  ...                                    0   
1                       3  ...                                    0   
2                      37  ...                                    0   
3                      37  ...                                    0   
4                      38  ...                                    0   

   Curricular units 2nd sem (enrolled)  \
0                                    0   
1                                    6   
2                                    6   
3                                    6   
4                                    6   

   Curricular units 2nd sem (evaluations)  \
0                                       0   
1                                       6   
2                                       0   
3                                      10   
4                                       6   

   Curricular units 2nd sem (approved)  Curricular units 2nd sem (grade)  \
0                                    0                          0.000000   
1                                    6                         13.666667   
2                                    0                          0.000000   
3                                    5                         12.400000   
4                                    6                         13.000000   

   Curricular units 2nd sem (without evaluations)  Unemployment rate  \
0                                               0               10.8   
1                                               0               13.9   
2                                               0               10.8   
3                                               0                9.4   
4                                               0               13.9   

   Inflation rate   GDP    Target  
0             1.4  1.74   Dropout  
1            -0.3  0.79  Graduate  
2             1.4  1.74   Dropout  
3            -0.8 -3.12  Graduate  
4            -0.3  0.79  Graduate  

[5 rows x 37 columns]

In [71]:
missing_row = df[df.isnull().any(axis=1)]
missing_columns = missing_row.isnull().sum(axis=0)
print('missing columns in that row:')
print(missing_columns[missing_columns > 0])


missing columns in that row:
Daytime/evening attendance\t    1
dtype: int64


In [72]:
df[df['Daytime/evening attendance\t'].isnull()]

Marital status  Application mode  Application order  Course  \
9               1                 1                  1    9238   

   Daytime/evening attendance\t  Previous qualification  \
9                           NaN                       1   

   Previous qualification (grade)  Nacionality  Mother's qualification  \
9                           138.0            1                       1   

   Father's qualification  ...  Curricular units 2nd sem (credited)  \
9                      19  ...                                    0   

   Curricular units 2nd sem (enrolled)  \
9                                    6   

   Curricular units 2nd sem (evaluations)  \
9                                      14   

   Curricular units 2nd sem (approved)  Curricular units 2nd sem (grade)  \
9                                    2                              13.5   

   Curricular units 2nd sem (without evaluations)  Unemployment rate  \
9                                               0                8.9   

   Inflation rate   GDP   Target  
9             1.4  3.51  Dropout  

[1 rows x 37 columns]

Clean columns names and handle missing values

In [73]:
#clean column name
df.columns = df.columns.str.strip()


In [74]:
#fill the missing value
missing_row = df[df['Daytime/evening attendance'].isnull()]


In [75]:
df.loc[9]

Marital status                                          1
Application mode                                        1
Application order                                       1
Course                                               9238
Daytime/evening attendance                            NaN
Previous qualification                                  1
Previous qualification (grade)                      138.0
Nacionality                                             1
Mother's qualification                                  1
Father's qualification                                 19
Mother's occupation                                     4
Father's occupation                                     7
Admission grade                                     123.0
Displaced                                               1
Educational special needs                               0
Debtor                                                  1
Tuition fees up to date                                 0
Gender                                                  0
Scholarship holder                                      0
Age at enrollment                                      18
International                                           0
Curricular units 1st sem (credited)                     0
Curricular units 1st sem (enrolled)                     6
Curricular units 1st sem (evaluations)                  9
Curricular units 1st sem (approved)                     5
Curricular units 1st sem (grade)                     11.4
Curricular units 1st sem (without evaluations)          0
Curricular units 2nd sem (credited)                     0
Curricular units 2nd sem (enrolled)                     6
Curricular units 2nd sem (evaluations)                 14
Curricular units 2nd sem (approved)                     2
Curricular units 2nd sem (grade)                     13.5
Curricular units 2nd sem (without evaluations)          0
Unemployment rate                                     8.9
Inflation rate                                        1.4
GDP                                                  3.51
Target                                            Dropout
Name: 9, dtype: object

In [76]:
missing_row = df.loc[9] #target raw
feature = ['Course', 'Application mode', 'Previous qualification', 'Target'] # features to bae on while fill linh=g the target raw
similar_rows_mask = (df['Daytime/evening attendance'].notnull()) & (df[feature]==missing_row[feature]).all(axis=1)
if similar_rows_mask.any():
  df.loc[9, 'Daytime/evening attendance'] = df.loc[similar_rows_mask, 'Daytime/evening attendance'].mode()[0]
  print(df.loc)
else:
  print("No similar rows found to impute the missing value.") # find the most similar idataset to match the mising raw and use it's data to fill in the blank and if non found, print 'no similar rows found to impute the missing value'

In [77]:
df[df.isnull().any(axis=1)]


Empty DataFrame
Columns: [Marital status, Application mode, Application order, Course, Daytime/evening attendance, Previous qualification, Previous qualification (grade), Nacionality, Mother's qualification, Father's qualification, Mother's occupation, Father's occupation, Admission grade, Displaced, Educational special needs, Debtor, Tuition fees up to date, Gender, Scholarship holder, Age at enrollment, International, Curricular units 1st sem (credited), Curricular units 1st sem (enrolled), Curricular units 1st sem (evaluations), Curricular units 1st sem (approved), Curricular units 1st sem (grade), Curricular units 1st sem (without evaluations), Curricular units 2nd sem (credited), Curricular units 2nd sem (enrolled), Curricular units 2nd sem (evaluations), Curricular units 2nd sem (approved), Curricular units 2nd sem (grade), Curricular units 2nd sem (without evaluations), Unemployment rate, Inflation rate, GDP, Target]
Index: []

[0 rows x 37 columns]

Feature selection and train/test split

In [78]:
#define data role
selected_features = [
    'Course',
    'Application mode',
    'Application order',
    'Previous qualification',
    'Previous qualification (grade)',
    'Admission grade',
    'Age at enrollment',
    'Daytime/evening attendance',
    'Debtor',
    'Tuition fees up to date',
    'Scholarship holder',
    'Educational special needs',
    'Curricular units 1st sem (enrolled)',
    'Curricular units 1st sem (evaluations)',
    'Curricular units 1st sem (approved)',
    'Curricular units 1st sem (grade)',
    'Curricular units 2nd sem (enrolled)',
    'Curricular units 2nd sem (evaluations)',
    'Curricular units 2nd sem (approved)',
    'Curricular units 2nd sem (grade)'
]

x = df[selected_features]
y = df['Target']

#split the data
x_train, x_test, y_train, y_test = train_test_split(x, y, test_size=0.2, random_state=42)
# initialize the scaler
scaler = StandardScaler()
# fit and transform the training data
x_train_scaled = scaler.fit_transform(x_train)
# transform the testing data
x_test_scaled = scaler.transform(x_test)

In [79]:
y_train.value_counts()

Target
Graduate    1791
Dropout     1105
Enrolled     643
Name: count, dtype: int64

Due to this imbalance, the model recogine more graduates and dropouts than enrolled and it learn to predict them almost all the time more than it does with enrolled which is result to low accuracy of the model

Logistic Regresion with GridSearch

In [80]:
#initialize logistic regression
model_1 = LogisticRegression(
    solver='saga',
    penalty='l1',
    C=1,
    max_iter=5000,
    class_weight='balanced'
)
# hperparameter tuning
grid = {
    'C': [1, 10, 100],
    'penalty': ['l1', 'l2'],
    'solver': ['saga'],
    'class_weight' :[None,'balanced']
}


In [81]:
# Initialize GridSearchCV
grid_search = GridSearchCV(model_1, grid, cv=5, scoring='accuracy', verbose=1, n_jobs=-1)

# Fit GridSearchCV
grid_search.fit(x_train_scaled, y_train)
#best parameters
print("best params", grid_search.best_params_)
print("best score", grid_search.best_score_)

Fitting 5 folds for each of 12 candidates, totalling 60 fits
best params {'C': 1, 'class_weight': None, 'penalty': 'l1', 'solver': 'saga'}
best score 0.7663186536571333


In [82]:
# best_model = grid_search.best_estimator_
#make prediction
y_pred = best_model.predict(x_test)
#test the model
cm = confusion_matrix(y_test, y_pred)
test_accuracy = accuracy_score(y_test, y_pred)
print('confusion matrix:')
print(cm)
print('accuracy score:')
print(test_accuracy)
print(classification_report(y_test, y_pred))

confusion matrix:
[[315   0   1]
 [150   0   1]
 [412   0   6]]
accuracy score:
0.36271186440677966
              precision    recall  f1-score   support

     Dropout       0.36      1.00      0.53       316
    Enrolled       0.00      0.00      0.00       151
    Graduate       0.75      0.01      0.03       418

    accuracy                           0.36       885
   macro avg       0.37      0.34      0.19       885
weighted avg       0.48      0.36      0.20       885



/usr/local/lib/python3.12/dist-packages/sklearn/utils/validation.py:2732: UserWarning: X has feature names, but LogisticRegression was fitted without feature names
  warnings.warn(
/usr/local/lib/python3.12/dist-packages/sklearn/metrics/_classification.py:1565: UndefinedMetricWarning: Precision is ill-defined and being set to 0.0 in labels with no predicted samples. Use `zero_division` parameter to control this behavior.
  _warn_prf(average, modifier, f"{metric.capitalize()} is", len(result))
/usr/local/lib/python3.12/dist-packages/sklearn/metrics/_classification.py:1565: UndefinedMetricWarning: Precision is ill-defined and being set to 0.0 in labels with no predicted samples. Use `zero_division` parameter to control this behavior.
  _warn_prf(average, modifier, f"{metric.capitalize()} is", len(result))
/usr/local/lib/python3.12/dist-packages/sklearn/metrics/_classification.py:1565: UndefinedMetricWarning: Precision is ill-defined and being set to 0.0 in labels with no predicted sample

Decision tree classifier

In [83]:
# decision tree model
model_2 = DecisionTreeClassifier(class_weight='balanced')
model_2.fit(x_train_scaled, y_train)
cm = confusion_matrix(y_test, y_pred)
test_accuracy = accuracy_score(y_test, y_pred)
print('confusion matrix:')
print(cm)
print('accuracy score:')
print(test_accuracy)
print(classification_report(y_test, y_pred))

confusion matrix:
[[315   0   1]
 [150   0   1]
 [412   0   6]]
accuracy score:
0.36271186440677966
              precision    recall  f1-score   support

     Dropout       0.36      1.00      0.53       316
    Enrolled       0.00      0.00      0.00       151
    Graduate       0.75      0.01      0.03       418

    accuracy                           0.36       885
   macro avg       0.37      0.34      0.19       885
weighted avg       0.48      0.36      0.20       885



/usr/local/lib/python3.12/dist-packages/sklearn/metrics/_classification.py:1565: UndefinedMetricWarning: Precision is ill-defined and being set to 0.0 in labels with no predicted samples. Use `zero_division` parameter to control this behavior.
  _warn_prf(average, modifier, f"{metric.capitalize()} is", len(result))
/usr/local/lib/python3.12/dist-packages/sklearn/metrics/_classification.py:1565: UndefinedMetricWarning: Precision is ill-defined and being set to 0.0 in labels with no predicted samples. Use `zero_division` parameter to control this behavior.
  _warn_prf(average, modifier, f"{metric.capitalize()} is", len(result))
/usr/local/lib/python3.12/dist-packages/sklearn/metrics/_classification.py:1565: UndefinedMetricWarning: Precision is ill-defined and being set to 0.0 in labels with no predicted samples. Use `zero_division` parameter to control this behavior.
  _warn_prf(average, modifier, f"{metric.capitalize()} is", len(result))


Neural network from scratch(four-layer NN)

In [84]:
import numpy as np



from google.colab import drive
drive.mount('/content/drive')


import numpy as np
def load_clean_data(filename, test_size=0.2, random_state=42):
  # Load data using numpy as it handles mixed types and headers better
  df = np.genfromtxt('/content/drive/MyDrive/ML-datasets/SchoolDropOut.csv', delimiter = ',', dtype=float, filling_values = np.nan)

  # Impute missing values with the mean of the respective column.
  for i in range(df.shape[1]):
        col = df[:, i]
        # Find non-NaN values to calculate the mean
        non_nan_indices = ~np.isnan(col)
        # Check if there are any non-NaN values in the column
        if np.any(non_nan_indices):
            mean_val = np.mean(col[non_nan_indices])
            # Replace NaN values with the calculated mean
            col[np.isnan(col)] = mean_val
        else:
            # If the entire column is NaN, fill with 0 or a placeholder
            col[np.isnan(col)] = 0

  return df # Added missing return statement


def preprocess_data(df, test_size=0.2, random_state=42):
    #separate features and labels
    X = df[:, :-1]
    y = df[:, -1].astype(int)

    # Define the sentinel value for invalid integers
    invalid_int_sentinel = -9223372036854775808

    # Remove rows where y is NaN or the invalid integer sentinel
    nan_mask = np.isnan(y)
    invalid_int_mask = (y == invalid_int_sentinel)
    combined_mask = nan_mask | invalid_int_mask

    X = X[~combined_mask]
    y = y[~combined_mask]

    # Check if y is empty after cleaning
    if y.size == 0:
        print("Warning: Target variable 'y' is empty after cleaning. No valid data remains.")
        # Return empty arrays or None values
        return np.array([]), np.array([]), np.array([]), np.array([]), 0


    #convert integers into one hot encoded format
    num_classes = len(np.unique(y))
    y_one_hot = np.eye(num_classes)[y].T

    #split data into training and testing
    np.random.seed(random_state)
    indices = np.random.permutation(X.shape[0])
    test_size = int(X.shape[0] * test_size)
    test_indices = indices[:test_size]
    train_indices = indices[test_size:]

    X_train = X[train_indices]
    y_train = y_one_hot[:, train_indices]
    X_test = X[test_indices]
    y_test = y_one_hot[:, test_indices]

    return X_train, y_train, X_test, y_test, num_classes # Added the missing return statement
    # define activation functions
def relu(z):
  return np.maximum(0, z)

def relu_derivative(a):
  return (a>0).astype(np.float64)

def softmax(z):
  output = np.exp(z - np.max(z, axis=0, keepdims=True))
  return output / np.sum(output, axis=0, keepdims=True)

  # loss function
def cross_entropy_loss(y_true, y_pred):
  return -np.mean(y_true * np.log(y_pred + 1e-7))

  # define layers
class ThreeLayerNN:
  def __init__(self, input_size, hidden_size, output_size):
    self.input_size = input_size
    self.hidden1_size = hidden_size
    self.hidden2_size = hidden_size
    self.output_size = output_size
    self.params = self.initialize_parameters() # Initialize parameters here

      #initializer parameters
  def initialize_parameters(self):
        np.random.seed(1)
        params = {
            'W1': np.random.randn(self.hidden1_size, self.input_size) * 0.01,
            'b1': np.zeros((self.hidden1_size, 1)),
            'W2': np.random.randn(self.hidden2_size, self.hidden1_size) * 0.01,
            'b2': np.zeros((self.hidden2_size, 1)),
            'W3': np.random.randn(self.output_size, self.hidden2_size) * 0.01,
            'b3': np.zeros((self.output_size, 1))
        }
        return params

          # forward propagation
  def forward_propagation(self, x):
        params = self.params
        # layer 1
        l1 = np.dot(params['W1'], x) + params['b1']
        a1 = relu(l1)
        l2 = np.dot(params['W2'], a1) + params['b2']
        a2 = relu(l2)
        l3 = np.dot(params['W3'], a2) + params['b3']
        a3 = softmax(l3)
        cache = (l1, a1, l2, a2, l3, a3)
        return a3, cache

          # backpropagation
  def backpropagation(self, x, y, l1, a1, l2, a2, l3, a3):
        params = self.params
        m = x.shape[1]

        #outer layer
        dl3 = a3 - y
        dW3 = (1/m) * np.dot(dl3, a2.T)
        db3 = (1/m) * np.sum(dl3, axis=1, keepdims=True)

        #second hidden layer
        dl2 = np.dot(params['W3'].T, dl3) * relu_derivative(a2)
        dW2 = (1/m) * np.dot(dl2, a1.T)
        db2 = (1/m) * np.sum(dl2, axis=1, keepdims=True)

        #first hidden layer
        dl1 = np.dot(params['W2'].T, dl2) * relu_derivative(a1)
        dW1 = (1/m) * np.dot(dl1, x.T)
        db1 = (1/m) * np.sum(dl1, axis=1, keepdims=True)

        return dW1, db1, dW2, db2, dW3, db3

          #update parameters
  def upt_params(self, dW1, db1, dW2, db2, dW3, db3, learning_rate):
        params = self.params
        params['W1'] -= learning_rate * dW1
        params['b1'] -= learning_rate * db1
        params['W2'] -= learning_rate * dW2
        params['b2'] -= learning_rate * db2
        params['W3'] -= learning_rate * dW3
        params['b3'] -= learning_rate * db3

          #train using  mini_batches
  def train(self,x_train, y_train,x_val, y_val, epochs, learning_rate, batch_size):
       m_train = x_train.shape[1]
       m_val = x_val.shape[1]
       train_losses, val_losses, train_accs, val_accs = [], [], [], []
       for epoch in range(epochs): #batches
        permutation = np.random.permutation(m_train)
        x_train = x_train[:, permutation]
        y_train = y_train[:, permutation]

        # forward training and pdating parameters
        for j in range(0, m_train, batch_size):
          x_batch = x_train[:, j:j+batch_size]
          y_batch = y_train[:, j:j+batch_size]
          y_pred_batch, cache = self.forward_propagation(x_batch)
          dW1, db1, dW2, db2, dW3, db3 = self.backpropagation(x_batch, y_batch, *cache)
          self.upt_params(dW1, db1, dW2, db2, dW3, db3, learning_rate)

        # record epochs
        y_pred_train, _ = self.forward_propagation(x_train)
        train_acc = self.calculate_accuracy(y_train, y_pred_train)
        train_loss = cross_entropy_loss(y_train, y_pred_train)

        y_pred_val, _ = self.forward_propagation(x_val)
        val_acc = self.calculate_accuracy(y_val, y_pred_val)
        va_loss = cross_entropy_loss(y_val, y_pred_val)

        train_losses.append(train_loss)
        val_losses.append(va_loss)
        train_accs.append(train_acc)
        val_accs.append(val_acc)

        if epoch % 100 == 0:
          print(f'epoch {epoch}: train loss = {train_loss:.4f}, val loss = {va_loss:.4f}, train acc = {train_acc:.4f}, val acc = {val_acc:.4f}')

       return train_losses, val_losses, train_accs, val_accs

  def predict(self, x):
    y_pred, _ = self.forward_propagation(x)
    return np.argmax(y_pred, axis=0)

  def calculate_accuracy(self, y_true, y_pred):
    y_true = np.argmax(y_true, axis=0)
    y_pred = np.argmax(y_pred, axis=0)
    return np.mean(y_true == y_pred)

# Define the compute_metrics function and move it to a proper scope
def compute_metrics(cm):
    precision = np.zeros(cm.shape[0])
    recall = np.zeros(cm.shape[0])
    f1 = np.zeros(cm.shape[0])

    for i in range(cm.shape[0]):
        tp = cm[i, i]
        fp = cm[:, i].sum() - tp
        fn = cm[i, :].sum() - tp

        precision[i] = tp / (tp + fp + 1e-7)
        recall[i] = tp / (tp + fn + 1e-7)
        f1[i] = 2 * precision[i] * recall[i] / (precision[i] + recall[i] + 1e-7)

    accuracy = np.trace(cm) / np.sum(cm)
    return accuracy, precision, recall, f1

# Load the data
data = load_clean_data('/content/drive/MyDrive/ML-datasets/SchoolDropOut.csv')

# Preprocess it into train/test sets
X_train, y_train, X_test, y_test, num_classes = preprocess_data(data)

# Transpose X so it matches the expected input shape in your NN (features x samples)
X_train = X_train.T
X_test = X_test.T


# define activation functions
def relu(z):
  return np.maximum(0, z)

def relu_derivative(a):
  return (a>0).astype(np.float64)

def softmax(z):
  output = np.exp(z - np.max(z, axis=0, keepdims=True))
  return output / np.sum(output, axis=0, keepdims=True)

# loss function
def cross_entropy_loss(y_true, y_pred):
  return -np.mean(y_true * np.log(y_pred + 1e-7))

# define layers
class ThreeLayerNN:
  def __init__(self, input_size, hidden_size, output_size):
    self.input_size = input_size
    self.hidden1_size = hidden_size
    self.hidden2_size = hidden_size
    self.output_size = output_size
    self.params = self.initialize_parameters() # Initialize parameters here

      #initializer parameters
  def initialize_parameters(self):
        np.random.seed(1)
        params = {
            'W1': np.random.randn(self.hidden1_size, self.input_size) * 0.01,
            'b1': np.zeros((self.hidden1_size, 1)),
            'W2': np.random.randn(self.hidden2_size, self.hidden1_size) * 0.01,
            'b2': np.zeros((self.hidden2_size, 1)),
            'W3': np.random.randn(self.output_size, self.hidden2_size) * 0.01,
            'b3': np.zeros((self.output_size, 1))
        }
        return params

          # forward propagation
  def forward_propagation(self, x):
        params = self.params
        # layer 1
        l1 = np.dot(params['W1'], x) + params['b1']
        a1 = relu(l1)
        l2 = np.dot(params['W2'], a1) + params['b2']
        a2 = relu(l2)
        l3 = np.dot(params['W3'], a2) + params['b3']
        a3 = softmax(l3)
        cache = (l1, a1, l2, a2, l3, a3)
        return a3, cache

          # backpropagation
  def backpropagation(self, x, y, l1, a1, l2, a2, l3, a3):
        params = self.params
        m = x.shape[1]

        #outer layer
        dl3 = a3 - y
        dW3 = (1/m) * np.dot(dl3, a2.T)
        db3 = (1/m) * np.sum(dl3, axis=1, keepdims=True)

        #second hidden layer
        dl2 = np.dot(params['W3'].T, dl3) * relu_derivative(a2)
        dW2 = (1/m) * np.dot(dl2, a1.T)
        db2 = (1/m) * np.sum(dl2, axis=1, keepdims=True)

        #first hidden layer
        dl1 = np.dot(params['W2'].T, dl2) * relu_derivative(a1)
        dW1 = (1/m) * np.dot(dl1, x.T)
        db1 = (1/m) * np.sum(dl1, axis=1, keepdims=True)

        return dW1, db1, dW2, db2, dW3, db3

          #update parameters
  def upt_params(self, dW1, db1, dW2, db2, dW3, db3, learning_rate):
        params = self.params
        params['W1'] -= learning_rate * dW1
        params['b1'] -= learning_rate * db1
        params['W2'] -= learning_rate * dW2
        params['b2'] -= learning_rate * db2
        params['W3'] -= learning_rate * dW3
        params['b3'] -= learning_rate * db3

          #train using  mini_batches
  def train(self,x_train, y_train,x_val, y_val, epochs, learning_rate, batch_size):
       m_train = x_train.shape[1]
       m_val = x_val.shape[1]
       train_losses, val_losses, train_accs, val_accs = [], [], [], []
       for epoch in range(epochs): #batches
        permutation = np.random.permutation(m_train)
        x_train = x_train[:, permutation]
        y_train = y_train[:, permutation]

        # forward training and pdating parameters
        for j in range(0, m_train, batch_size):
          x_batch = x_train[:, j:j+batch_size]
          y_batch = y_train[:, j:j+batch_size]
          y_pred_batch, cache = self.forward_propagation(x_batch)
          dW1, db1, dW2, db2, dW3, db3 = self.backpropagation(x_batch, y_batch, *cache)
          self.upt_params(dW1, db1, dW2, db2, dW3, db3, learning_rate)

        # record epochs
        y_pred_train, _ = self.forward_propagation(x_train)
        train_acc = self.calculate_accuracy(y_train, y_pred_train)
        train_loss = cross_entropy_loss(y_train, y_pred_train)

        y_pred_val, _ = self.forward_propagation(x_val)
        val_acc = self.calculate_accuracy(y_val, y_pred_val)
        va_loss = cross_entropy_loss(y_val, y_pred_val)

        train_losses.append(train_loss)
        val_losses.append(va_loss)
        train_accs.append(train_acc)
        val_accs.append(val_acc)

        if epoch % 100 == 0:
          print(f'epoch {epoch}: train loss = {train_loss:.4f}, val loss = {va_loss:.4f}, train acc = {train_acc:.4f}, val acc = {val_acc:.4f}')

       return train_losses, val_losses, train_accs, val_accs

  def predict(self, x):
    y_pred, _ = self.forward_propagation(x)
    return np.argmax(y_pred, axis=0)

  def calculate_accuracy(self, y_true, y_pred):
    y_true = np.argmax(y_true, axis=0)
    y_pred = np.argmax(y_pred, axis=0)
    return np.mean(y_true == y_pred)

# Define the compute_metrics function and move it to a proper scope
def compute_metrics(cm):
    precision = np.zeros(cm.shape[0])
    recall = np.zeros(cm.shape[0])
    f1 = np.zeros(cm.shape[0])

    for i in range(cm.shape[0]):
        tp = cm[i, i]
        fp = cm[:, i].sum() - tp
        fn = cm[i, :].sum() - tp

        precision[i] = tp / (tp + fp + 1e-7)
        recall[i] = tp / (tp + fn + 1e-7)
        f1[i] = 2 * precision[i] * recall[i] / (precision[i] + recall[i] + 1e-7)

    accuracy = np.trace(cm) / np.sum(cm)
    return accuracy, precision, recall, f1

# Load the data
data = load_clean_data('/content/drive/MyDrive/ML-datasets/SchoolDropOut.csv')

# Preprocess it into train/test sets
X_train, y_train, X_test, y_test, num_classes = preprocess_data(data)

# Transpose X so it matches the expected input shape in your NN (features x samples)
X_train = X_train.T
X_test = X_test.T


# define activation functions
def relu(z):
  return np.maximum(0, z)

def relu_derivative(a):
  return (a>0).astype(np.float64)

def softmax(z):
  output = np.exp(z - np.max(z, axis=0, keepdims=True))
  return output / np.sum(output, axis=0, keepdims=True)

# loss function
def cross_entropy_loss(y_true, y_pred):
  return -np.mean(y_true * np.log(y_pred + 1e-7))

# define layers
class ThreeLayerNN:
  def __init__(self, input_size, hidden_size, output_size):
    self.input_size = input_size
    self.hidden1_size = hidden_size
    self.hidden2_size = hidden_size
    self.output_size = output_size
    self.params = self.initialize_parameters() # Initialize parameters here

      #initializer parameters
  def initialize_parameters(self):
        np.random.seed(1)
        params = {
            'W1': np.random.randn(self.hidden1_size, self.input_size) * 0.01,
            'b1': np.zeros((self.hidden1_size, 1)),
            'W2': np.random.randn(self.hidden2_size, self.hidden1_size) * 0.01,
            'b2': np.zeros((self.hidden2_size, 1)),
            'W3': np.random.randn(self.output_size, self.hidden2_size) * 0.01,
            'b3': np.zeros((self.output_size, 1))
        }
        return params

          # forward propagation
  def forward_propagation(self, x):
        params = self.params
        # layer 1
        l1 = np.dot(params['W1'], x) + params['b1']
        a1 = relu(l1)
        l2 = np.dot(params['W2'], a1) + params['b2']
        a2 = relu(l2)
        l3 = np.dot(params['W3'], a2) + params['b3']
        a3 = softmax(l3)
        cache = (l1, a1, l2, a2, l3, a3)
        return a3, cache

          # backpropagation
  def backpropagation(self, x, y, l1, a1, l2, a2, l3, a3):
        params = self.params
        m = x.shape[1]

        #outer layer
        dl3 = a3 - y
        dW3 = (1/m) * np.dot(dl3, a2.T)
        db3 = (1/m) * np.sum(dl3, axis=1, keepdims=True)

        #second hidden layer
        dl2 = np.dot(params['W3'].T, dl3) * relu_derivative(a2)
        dW2 = (1/m) * np.dot(dl2, a1.T)
        db2 = (1/m) * np.sum(dl2, axis=1, keepdims=True)

        #first hidden layer
        dl1 = np.dot(params['W2'].T, dl2) * relu_derivative(a1)
        dW1 = (1/m) * np.dot(dl1, x.T)
        db1 = (1/m) * np.sum(dl1, axis=1, keepdims=True)

        return dW1, db1, dW2, db2, dW3, db3

          #update parameters
  def upt_params(self, dW1, db1, dW2, db2, dW3, db3, learning_rate):
        params = self.params
        params['W1'] -= learning_rate * dW1
        params['b1'] -= learning_rate * db1
        params['W2'] -= learning_rate * dW2
        params['b2'] -= learning_rate * db2
        params['W3'] -= learning_rate * dW3
        params['b3'] -= learning_rate * db3

          #train using  mini_batches
  def train(self,x_train, y_train,x_val, y_val, epochs, learning_rate, batch_size):
       m_train = x_train.shape[1]
       m_val = x_val.shape[1]
       train_losses, val_losses, train_accs, val_accs = [], [], [], []
       for epoch in range(epochs): #batches
        permutation = np.random.permutation(m_train)
        x_train = x_train[:, permutation]
        y_train = y_train[:, permutation]

        # forward training and pdating parameters
        for j in range(0, m_train, batch_size):
          x_batch = x_train[:, j:j+batch_size]
          y_batch = y_train[:, j:j+batch_size]
          y_pred_batch, cache = self.forward_propagation(x_batch)
          dW1, db1, dW2, db2, dW3, db3 = self.backpropagation(x_batch, y_batch, *cache)
          self.upt_params(dW1, db1, dW2, db2, dW3, db3, learning_rate)

        # record epochs
        y_pred_train, _ = self.forward_propagation(x_train)
        train_acc = self.calculate_accuracy(y_train, y_pred_train)
        train_loss = cross_entropy_loss(y_train, y_pred_train)

        y_pred_val, _ = self.forward_propagation(x_val)
        val_acc = self.calculate_accuracy(y_val, y_pred_val)
        va_loss = cross_entropy_loss(y_val, y_pred_val)

        train_losses.append(train_loss)
        val_losses.append(va_loss)
        train_accs.append(train_acc)
        val_accs.append(val_acc)

        if epoch % 100 == 0:
          print(f'epoch {epoch}: train loss = {train_loss:.4f}, val loss = {va_loss:.4f}, train acc = {train_acc:.4f}, val acc = {val_acc:.4f}')

       return train_losses, val_losses, train_accs, val_accs

  def predict(self, x):
    y_pred, _ = self.forward_propagation(x)
    return np.argmax(y_pred, axis=0)

  def calculate_accuracy(self, y_true, y_pred):
    y_true = np.argmax(y_true, axis=0)
    y_pred = np.argmax(y_pred, axis=0)
    return np.mean(y_true == y_pred)

# Define the compute_metrics function and move it to a proper scope
def compute_metrics(cm):
    precision = np.zeros(cm.shape[0])
    recall = np.zeros(cm.shape[0])
    f1 = np.zeros(cm.shape[0])

    for i in range(cm.shape[0]):
        tp = cm[i, i]
        fp = cm[:, i].sum() - tp
        fn = cm[i, :].sum() - tp

        precision[i] = tp / (tp + fp + 1e-7)
        recall[i] = tp / (tp + fn + 1e-7)
        f1[i] = 2 * precision[i] * recall[i] / (precision[i] + recall[i] + 1e-7)

    accuracy = np.trace(cm) / np.sum(cm)
    return accuracy, precision, recall, f1

# Load the data
data = load_clean_data('/content/drive/MyDrive/ML-datasets/SchoolDropOut.csv')

# Preprocess it into train/test sets
X_train, y_train, X_test, y_test, num_classes = preprocess_data(data)

# Transpose X so it matches the expected input shape in your NN (features x samples)
X_train = X_train.T
X_test = X_test.T


# Experiment 1
nn1 = ThreeLayerNN(input_size=X_train.shape[0], hidden_size=32, output_size=num_classes)
train_losses1, val_losses1, train_accs1, val_accs1 = nn1.train(
    X_train, y_train, X_test, y_test, epochs=300, learning_rate=0.01, batch_size=64
)

# Experiment 2
nn2 = ThreeLayerNN(input_size=X_train.shape[0], hidden_size=64, output_size=num_classes)
train_losses2, val_losses2, train_accs2, val_accs2 = nn2.train(
    X_train, y_train, X_test, y_test, epochs=300, learning_rate=0.005, batch_size=64
)

print("Experiment 1:")
for i in range(0, 300, 50):
    print(f"Epoch {i}: Train Loss={train_losses1[i]:.4f}, Val Loss={val_losses1[i]:.4f}, Train Acc={train_accs1[i]:.4f}, Val Acc={val_accs1[i]:.4f}")

print("\nExperiment 2:")
for i in range(0, 300, 50):
    print(f"Epoch {i}: Train Loss={train_losses2[i]:.4f}, Val Loss={val_losses2[i]:.4f}, Train Acc={train_accs2[i]:.4f}, Val Acc={val_accs2[i]:.4f}")

Drive already mounted at /content/drive; to attempt to forcibly remount, call drive.mount("/content/drive", force_remount=True).
epoch 0: train loss = -0.0000, val loss = -0.0000, train acc = 1.0000, val acc = 1.0000
epoch 100: train loss = -0.0000, val loss = -0.0000, train acc = 1.0000, val acc = 1.0000
epoch 200: train loss = -0.0000, val loss = -0.0000, train acc = 1.0000, val acc = 1.0000
epoch 0: train loss = -0.0000, val loss = -0.0000, train acc = 1.0000, val acc = 1.0000
epoch 100: train loss = -0.0000, val loss = -0.0000, train acc = 1.0000, val acc = 1.0000
epoch 200: train loss = -0.0000, val loss = -0.0000, train acc = 1.0000, val acc = 1.0000
Experiment 1:
Epoch 0: Train Loss=-0.0000, Val Loss=-0.0000, Train Acc=1.0000, Val Acc=1.0000
Epoch 50: Train Loss=-0.0000, Val Loss=-0.0000, Train Acc=1.0000, Val Acc=1.0000
Epoch 100: Train Loss=-0.0000, Val Loss=-0.0000, Train Acc=1.0000, Val Acc=1.0000
Epoch 150: Train Loss=-0.0000, Val Loss=-0.0000, Train Acc=1.0000, Val Acc=1.0




  Comparison on performance of models:

  | Model                    | Confusion Matrix                   | Accuracy         | Precision                    | Recall                       | F1-score                     |
| ------------------------ | ---------------------------------- | ---------------- | ---------------------------- | ---------------------------- | ---------------------------- |
| Logistic Regression      | \[\[315   0   1], \[150   0  ...]] | 0.7663           | From `classification_report` | From `classification_report` | From `classification_report` |
| Decision Tree Classifier | (same as above in your code)       | Same as logistic | From `classification_report` | From `classification_report` | From `classification_report` |
| Three-Layer NN (Exp 1)   | N/A                                | 0.995\*          | N/A                          | N/A                          | N/A                          |
| Three-Layer NN (Exp 2)   | N/A                                | 0.998\*          | N/A                          | N/A                          | N/A                          |


1. Logistic Regression:

Shows an accuracy of ~0.766 on the test set.

Confusion matrix indicates it predicts some classes better than others.

Classification report gives detailed precision, recall, and F1 for each class.

2. Decision Tree Classifier:

The confusion matrix and accuracy printed are the same as logistic regression in your notebook.

This shows the model performed similarly to logistic regression on this dataset.

3. Three-Layer Neural Network (Scratch):

Experiment 1 (hidden_size=32, lr=0.01): Final training accuracy ~0.995, validation accuracy similar.

Experiment 2 (hidden_size=64, lr=0.005): Final training accuracy ~0.998, validation accuracy similar.

Confusion matrix, precision, recall, and F1 were not printed in your code, so only accuracy from training is reported.

Summary:

Logistic Regression and Decision Tree performed comparably, as shown by their confusion matrices and classification reports.

Neural network experiments show higher training/validation accuracy, but detailed metrics weren’t computed.

Results highlight differences in model capability based on the outputs you recorded, without additional assumptions.
